### DISCUSSION TITLE
TTree performance improvements advice needed

### DISCUSSION THREAD

**Discussion by JacekHoleczek** [scikit-hep/uproot5#1106] [Q&A] [answered]
I am trying to write a small python macro that should "demonstrate" that uproot is (almost) as fast as C++.
Could you please give me advice on how to improve it?

How can one make this code run even "faster"?

Can one somehow set the "title" of the created hist.Hist?
If not, can one somehow set the "title" of the generated TH1D histogram (when writing the hist.Hist to a ROOT file)?

In a second step later ...
I assume that each time I ask for some "array", uproot will fully load all TTree entries corresponding to the requested branch into RAM. What would be the easiest way to change this code so that the RAM usage was kept "small" (but still running with good performance), even when analyzing giant trees (i.e., with very many entries)?

```python
import uproot
import numpy as np
import hist

def process():
        # open the ROOT file and load the hk TTree
        with uproot.open({'data/my_tree.root': 'hk'}) as t:
                # test that we can read some available arrays
                tq_real = t.arrays(['TQReal.nhits', 'TQReal.pc2pe', 'TQReal.t0'])
                tq_real_hits = t.arrays(['TQReal.hits.cable', 'TQReal.hits.T', 'TQReal.hits.Q'])
                n_high_hit = np.sum(tq_real_hits['TQReal.hits.Q'] > 1.0)
                print(f'Number of high Q hits: {n_high_hit}')
                # create and fill some histograms
                h_goodness = hist.Hist(hist.axis.Regular(100, 0.0, 1.0, label = 'goodness_time_fit (RecoBonsai)'))
                h_goodness.fill(t['RecoBonsai.goodness_time_fit'].array())
                h_trigger = hist.Hist(hist.axis.Regular(10, 0.0, 10.0, label = 'trigger_id (Header)'))
                h_trigger.fill(t['Header.trigger_id'].array())
                h_goodness_leaf = None
                if 'RecoLEAF.goodness' in t: # RecoLEAF is a new class, not available everywhere
                        h_goodness_leaf = hist.Hist(hist.axis.Regular(100, 0.0, 1.0, label = 'goodness (RecoLEAF)'))
                        h_goodness_leaf.fill(t['RecoLEAF.goodness'].array())
                # write histograms to a ROOT file
                # note: when a new ROOT TH[123]D is created from the
                #       corresponding hist.Hist below, its "statistics"
                #       is computed from its bin content (in bare ROOT,
                #       the "statistics" is computed at filling time)
                with uproot.recreate('data/output_analysis_tree_python_v1.root') as f_out:
                        f_out['goodness'] = h_goodness
                        f_out['trigger'] = h_trigger
                        if h_goodness_leaf is not None:
                                f_out['goodness_leaf'] = h_goodness_leaf

if __name__ == '__main__':
        print('Processing')
        process()
        print('Done')
```

**Comment by jpivarski** [comment, marked as answer]
If these arrays are not ragged (i.e. `t[branchname].typename` is a C++ value type, such as `int32_t` or `double`, possibly fixed-size arrays like `float[10]`, but not variable-length arrays like `float[]`), then you can use NumPy with `library="np"` instead of the default Awkward Array, and that would have greater or equal speed.

Be careful to avoid reading the same data multiple times. I don't see any examples of that here, but you can either consolidate all of your array-reading functions in one [uproot.TTree.arrays](https://uproot.readthedocs.io/en/latest/uproot.behaviors.TTree.TTree.html#arrays) call or use the `array_cache` argument in both [uproot.TTree.arrays](https://uproot.readthedocs.io/en/latest/uproot.behaviors.TTree.TTree.html#arrays) and [uproot.TBranch.array](https://uproot.readthedocs.io/en/latest/uproot.behaviors.TBranch.TBranch.html#array). (It doesn't do anything magical; it just checks a MutableMapping (dict) that you maintain, first, instead of reading from the file.)

If your data are heavily compressed with LZMA, there might be some benefit from setting a `decompression_executor` (give it a [concurrent.futures.ThreadPoolExecutor](https://docs.python.org/3/library/concurrent.futures.html#threadpoolexecutor)). I'm including the caveats (and not recommending `interpretation_executor`) because Python's GIL prevents it from taking advantage of multiple threads in the same process. Decompression, however, is offloaded to external libraries that release the GIL, so we have seen some improvements from multithreading just that part.

If you gather all of the array-fetching into one call, it would be easier to convert this into a process that calls [uproot.TTree.iterate](https://uproot.readthedocs.io/en/latest/uproot.behaviors.TTree.TTree.html#iterate) to read a subset of entries in each iteration step. That would allow you to scale to files that are too big to load into memory all at once (while still being a single-threaded process). The amount of data read in each step is controlled by `step_size`, which should be as large as you can make it while not running out of memory. The iteration steps don't need to line up with the TBaskets (granularity in the ROOT file itself), and any mismatches from one step of iteration are carried over to the next step of iteration (so it's a little better than using `entry_start` and `entry_stop` to cut up the file manually).

To convert `process` into a task that deals with a subset of entries at a time, you'll either need to collect and add your histograms (with predefined binning, as you have them) or define empty histograms outside of the [uproot.TTree.iterate](https://uproot.readthedocs.io/en/latest/uproot.behaviors.TTree.TTree.html#iterate) or [uproot.iterate](https://uproot.readthedocs.io/en/latest/uproot.behaviors.TBranch.iterate.html) loop and `fill` them repeatedly. hist's `fill` function can be used for multiple array batches.

Beyond that, you want to consider parallel-processing. Because of Python's GIL, that almost always means [multiprocessing](https://docs.python.org/3/library/multiprocessing.html), which, in turn, means not sending too much data between the processes because it has to be serialized (can't be just referenced as threads can). Your histograms are probably small enough. Setting this manually is annoying, so consider [uproot.dask](https://uproot.readthedocs.io/en/latest/uproot._dask.dask.html) to distribute them on a [Dask](https://www.dask.org/) cluster. ([dask-histogram](https://github.com/dask-contrib/dask-histogram) is a Daskified version of boost-histogram/hist.)

(How you set up or access the Dask cluster depends on where you are. In [tutorials like this one](https://github.com/jpivarski-talks/2023-07-24-tac-hep-tutorial/blob/main/horizontal.ipynb), we launched a few Dask workers on one computer to show the principle. The [Coffea](https://github.com/CoffeaTeam/coffea) team has extensive experience with Dask, and you might want to consider using Coffea to get some features built-in.

------------------------

The issues involved in horizontal scaling applies to C++ and Python equally, and Python may have an advantage there because of its simplicity—much of the difficulty of horizontal scaling comes from complexity, and simplicity in the workload and interfaces lets you concentrate on the problems in horizontal scaling.

For vertical scaling, however, Python can only approach the speed of C++ in specific contexts, when it "gets out of the way." Uproot takes advantage of a fact that is true of some but not all ROOT datasets: pure numerical and ragged-numerical (only one level deep) data are stored in the ROOT file as arrays, so we can just cast those arrays instead of iterating over them. (The equivalent in C++ ROOT is called [Bulk I/O](https://arxiv.org/abs/1906.08169).) Uproot can only approach ROOT's read speed to the extent that this is true: if the individual TBaskets are large (10's or 100's of kB at least) and the data types are right. In that case, reading is dominated by the physical disk speed and decompression, which are the same for Python and C++. In the worst case, you have lots of small files or lots of small TBaskets in them and Python/Uproot will spend most of its time deserializing this metadata. (This worst case is bad for C++, too, but it's proportionally worse for Python.)

If you want to go even further afield, you could consider [UnROOT.jl](https://github.com/JuliaHEP/UnROOT.jl) to load the data in Julia. Right now, that means also doing the analysis in Julia, but we're working on ways ([AwkwardArray.jl](https://github.com/JuliaHEP/AwkwardArray.jl)) to connect the two environments.

**Comment by JacekHoleczek** [comment]
What about the "title" of the created hist.Hist?
Can one somehow set it or later set the "title" of the generated TH1D histogram (when writing the hist.Hist to a ROOT file)?

**Reply by jpivarski** [reply]
The `title` is a property of the histogram object, and in Uproot, the `name` is set when you assign it into a ROOT file, like

```python
output_file[name] = hist.Hist(... title ...)
```

In ROOT, both `fTitle` and `fName` are attributes of the object, but then it has to ensure that the `fName` in the object agrees with the look-up name in the TDirectory, and the look-up names have to be unique apart from cycle number—that's why I went with the policy of Uproot assigning a name at exactly the time when it's put into a directory.

--------------------

I had thought that one of the hist metadata attributes is interpreted as a title...

https://github.com/scikit-hep/uproot5/blob/eaa3032ea271f1d468f6af08ac2f14e8cd2e0fcb/src/uproot/writing/identify.py#L259-L261

But it looks like Uproot takes the output histogram title from any `title` or `name` attribute on the histogram object itself. But the good news is that you can assign any attribute to a hist object:

```python
>>> h = hist.Hist.new.Reg(10, -5, 5).Double()
>>> h.title = "meow"
>>> h.title
'meow'
```

(It's not a dataclass or restricted with `__slots__` in such a way that would prevent you from doing that.)

Just be careful to set the title of the specific histogram that's getting written. If you're filling histograms in parallel, you'll have many identical-apart-from-bin-contents histograms, an arbitrarily assigned attribute like `title` wouldn't be copied when they get merged together.

**Reply by JacekHoleczek** [reply]
Well, I am getting an error (`'Hist' object has no attribute 'title'`) from:
`h = hist.Hist(hist.axis.Regular(10, -5.0, 5.0)); h.title("meow")`

I am asking about the "title", because in ROOT one can use (up to 3 axes): `h->SetTitle("Histo title;X title;Y title;Z title");`
With "hist.Hist", when I create a "hist.axis.Regular", I can use: "`label='X title'`"
I have no way to set the "`Y title`" (because I do not create the second axis).

BTW. In ROOT, your histogram can have one name in RAM, and then another name can be set when writing it to a file (but the name in RAM will not be changed), e.g.: `h->SetName("SomeName"); h->Write("AnotherName");`

**Reply by jpivarski** [reply]
> Well, I am getting an error (`'Hist' object has no attribute 'title'`) from:
> `h = hist.Hist(hist.axis.Regular(10, -5.0, 5.0)); h.title("meow")`

It's

```python
h.title = "meow"
```

not

```python
h.title("meow")
```

-----------------

I didn't know about the semicolon-parsing in `h->SetTitle("title;x;y;z")`. It must be an active mechanism to pass titles down to each TAxis. Checking the Uproot code again, it looks like there's a way to get that from hist into ROOT:

https://github.com/scikit-hep/uproot5/blob/eaa3032ea271f1d468f6af08ac2f14e8cd2e0fcb/src/uproot/writing/identify.py#L331-L342

It's an `axis` attribute named `label` or `name`. Note that that's on the histogram's `axis`, not on the histogram itself (the distinction that ROOT's semicolon-notation is making).

```python
>>> h = hist.Hist(hist.axis.Regular(10, -5.0, 5.0, label="meow"))
>>> h.axes[0].label
'meow'
```

but

```python
>>> h.label is None
True
```

Each of hist's axis type constructors has a `label` and a `name` constructor argument ([documented here](https://hist.readthedocs.io/en/latest/user-guide/axes.html)). I don't think there's a preferred convention for whether you should use "label" or "name".

----------

> BTW. In ROOT, your histogram can have one name in RAM, and then another name can be set when writing it to a file (but the name in RAM will not be changed), e.g.: `h->SetName("SomeName"); h->Write("AnotherName");`

Thanks, that's good to know! I was just saying why we decided to not even have names on the objects in Uproot, so that there would only be one source of truth.

**Reply by JacekHoleczek** [reply]
Many thanks ... I did this mistake also in my "test script" ... yes, it needs to be `h.title = "meow"`

I tried:
`h_trigger.title = 'Histo title;X title;Y title'`
and then the histogram's title is the whole string (ROOT would strip everything starting from the first "`;`" character) and the axes titles are not changed.
Not a problem for me.
So, I now know how to set "Histo title"  and the "X title" (using the "`label ='X title'`" argument).
What I still miss is how to set the "Y title" (for a 1D histogram, I explicitly create the X axis only).

**Reply by jpivarski** [reply]
These things are accessible if you're willing to convert from hist to TH* manually. The code in [uproot/writing/identify.py](https://github.com/scikit-hep/uproot5/blob/main/src/uproot/writing/identify.py) recognizes hist objects and builds the corresponding TH* using [to_TAxis](https://uproot.readthedocs.io/en/latest/uproot.writing.identify.to_TAxis.html), [to_TH1x](https://uproot.readthedocs.io/en/latest/uproot.writing.identify.to_TH1x.html), etc. functions. If you do the same, to make an Uproot TH* instance, Uproot will recognize it when you assign it into a file with

```python
output_file[name] = some_thistogram_object
```

and not modify the titles that you assign to it.

**Reply by JacekHoleczek** [reply]
O.K. That would be too complicated for a simple "demonstrate" script.

I've got a strange question ... with Uproot 5.0.13, my small script needed 0.7s. Now, with Uproot 5.2.2 it needs 1.0s (same Python 3.11.7 in a conda-forge "scikit-hep" environment in both cases). A compiled C++ application needs 0.6s.

UPDATE (2024.01.30): I created a test TTree with 100 times more entries (900MB ROOT file), and then Uproot 5.0.13 needed 9.1s while Uproot 5.2.2 needed 10.1s. A compiled C++ application needed 8.6s. Not bad, don't you think?

**Reply by jpivarski** [reply]
Between 5.1.x and 5.2.x, the low-level backend changed from memory-mapping to fsspec. A rate, particularly for such a short time-interval test, could change at the 30% level with differences like that. (Although it's worse in this case, the same change might be better for other cases, the only way to make robust conclusions is to test changes in isolation and focus on large-limit scaling, and those tests were performed before committing to fsspec.)

**Reply by JacekHoleczek** [reply]
Would it make sense to raise the "(n+1)-axis" title problem on the "hist" (or "UHI") forum?
When one creates an "n-dimensional" histogram, one can easily set the titles of all "n-axes" (which are explicitly created). But it seems there is no way to set the title of the "(n+1)-axis", which is then shown when drawing the histogram.
Maybe they could add an additional attribute (?), like for the additional global histogram "title".

**Reply by jpivarski** [reply]
It's definitely a hist thing, not an Uproot thing, so that's where it should be discussed.

Is this the "number of counts" or "sum of weights per bin" axis? That's not an ordinary histogram axis, and ROOT doesn't have a TAxis corresponding to that axis, either. I can see why you want to have a place to put a label like "counts in 1/GeV", but I don't know where it would go.

**Reply by JacekHoleczek** [reply]
Actually, ROOT keeps the information about all 3 axes (even if some are not needed/used):
```cpp
TH1F *h = new TH1F("h", "MyH;MyX;MyY;MyZ", 1, 0., 1.);
std::cout << h->GetName() << " " << h->GetXaxis()->GetName() << " " << h->GetYaxis()->GetName() << " " << h->GetZaxis()->GetName() << std::endl;
std::cout << h->GetTitle() << " " << h->GetXaxis()->GetTitle() << " " << h->GetYaxis()->GetTitle() << " " << h->GetZaxis()->GetTitle() << std::endl;
```

Where it could go? Well, we have now the "title" attribute (which sets the global histogram title), so I could imagine an additional "label" attribute which would then set the "(n+1)-axis" title.

**Reply by jpivarski** [reply]
Those three are binning axes, they specify how many bins and what range, which are not relevant for counts/sums of bin weights. In a 3D histogram, where would the "counts/GeV" label go?

I agree that we (or the hist/UHI developers) can make a new attribute to label this counts-dimension (not axis, or anyway, not a binning-axis). Maybe it has already been done and I don't know what its name is. When I plot things with UHI, the plot has the word "counts" on that plot axis, and maybe that's a default if it doesn't find something in one of the histogram's own attributes.

**Reply by JacekHoleczek** [reply]
In principle, for me, the "(n+1)-axis" title describes the "physical meaning" of the "content" of bins (e.g., "number of events" or "counts/GeV").
The drawing library then can decide what happens to the "(n+1)-axis" title.
Some drawing tools may use it, some may not.
For a `TH1`, this would be the drawn title of the drawn y-axis.
For a `TH2`, if you use the "`Z`" drawing option, an additional "color palette" is drawn, and it will get its title from the histogram's "3rd axis" title (if the user sets it, of course). Without the "`Z`" drawing option, it will simply be the drawn title of the drawn z-axis.
The problem with a `TH3` is that ROOT keeps 3 axes only, so when drawing, you would need to manually set the title of the "color palette" (if you wanted it). I don't think a `TH3` can keep the "4th axis" title (at least, I can't remember how to do it).

**Reply by JacekHoleczek** [reply]
Extremely hot (watch out for your fingertips when you click it): "[`[Z (bad) label for palette axis in TH3D`](https://root-forum.cern.ch/t/z-bad-label-for-palette-axis-in-th3d/53317/18)"

**Comment by JacekHoleczek** [comment]
Many, many thanks for your help and all the additional info (I'll remember what you wrote about the scaling to the "iterative" / "parallel-processing" approach).

Here's my final version for the time being:

```python
import uproot
import numpy as np
import hist

def process():
        # open the ROOT file and load the hk TTree
        with uproot.open({'data/my_tree.root': 'hk'}) as t:
                # test that we can read some available branches
                # note: if branches keep C++ fundamental data types or their fixed-size arrays,
                #       one can use the library='np' option (the NumPy, which may improve speed)
                tq_real = t.arrays(['TQReal.nhits', 'TQReal.pc2pe', 'TQReal.t0'], library='np')
                # note: if any branches with ragged/variable-length arrays are retrieved,
                #       one must use the default library='ak' option (the Awkward Array)
                tq_real_hits = t.arrays(['TQReal.hits.cable', 'TQReal.hits.T', 'TQReal.hits.Q'])
                n_high_hit = np.sum(tq_real_hits['TQReal.hits.Q'] > 1.0)
                print(f'Number of high Q hits: {n_high_hit}')
                # create and fill some histograms
                h_goodness = hist.Hist(hist.axis.Regular(100, 0.0, 1.0, label='goodness_time_fit (RecoBonsai)'))
                h_goodness.title = 'bs goodness distribution'
                h_goodness.fill(t['RecoBonsai.goodness_time_fit'].array(library='np'))
                h_trigger = hist.Hist(hist.axis.Regular(10, 0.0, 10.0, label='trigger_id (Header)'))
                h_trigger.title = 'trigger distribution'
                h_trigger.fill(t['Header.trigger_id'].array(library='np'))
                h_goodness_leaf = None
                if 'RecoLEAF.goodness' in t: # RecoLEAF is a new class, not available everywhere
                        h_goodness_leaf = hist.Hist(hist.axis.Regular(100, 0.0, 1.0, label='goodness (RecoLEAF)'))
                        h_goodness_leaf.title = 'leaf goodness distribution'
                        h_goodness_leaf.fill(t['RecoLEAF.goodness'].array(library='np'))
                # write histograms to a ROOT file
                # note: when a new ROOT TH[123]D is created from the
                #       corresponding hist.Hist below, its "statistics"
                #       is computed from its bin content (in bare ROOT,
                #       the "statistics" is computed at filling time)
                with uproot.recreate('data/output_analysis_tree_python_v1.root') as f_out:
                        f_out['goodness'] = h_goodness
                        f_out['trigger'] = h_trigger
                        if h_goodness_leaf is not None:
                                f_out['goodness_leaf'] = h_goodness_leaf

if __name__ == '__main__':
        print('Processing')
        process()
        print('Done')
```
